File: src/game/game.py
from __future__ import annotations

from time import sleep
from datetime import datetime
import math
from typing import List

BOARD_WIDTH = 16

board = ["X" for x in range(BOARD_WIDTH * BOARD_WIDTH)]


def quad(a, b, c):
    determinant = b * b - 4 * a * c
    # Avoid math domain error. Since we only care about collisions that happen between t=[0, 1)
    # it is ok to return -1 here
    if determinant < 0:
        return -1
    else:
        # Similarly we only care about the first touching so we don't have to care about the
        # positive root
        return (-b - math.sqrt(determinant)) / (2 * a)


def collision(start1, v1, start2, v2, radius=1):
    # solves the following equation (x'(0) - x'(1))^2 + (y'(0) - y'(1))^2 = radius ^ 2 regards to time
    # Basically when this equation equals to zero the two objects touch.
    a = (v1[0] - v2[0]) ** 2 + (v1[1] - v2[1]) ** 2
    b = -2 * (v1[0] - v2[0]) * (start2[0] - start1[0]) - 2 * (v1[1] - v2[1]) * (start2[1] - start1[1])
    c = (start2[0] - start1[0]) ** 2 + (start2[1] - start1[1]) ** 2 - radius ** 2
    if a == 0 and b:
        return -c / b
    if a == 0 and b == 0:
        return -1
    return quad(a, b, c)


class MovingObject:
    deceleration = -0.1
    BOARD_INSIDE = BOARD_WIDTH - 3

    def __init__(self, starting_velocity=(0, 0), symbol="*"):
        self.velocity = list(starting_velocity)
        self.location = list((self.BOARD_INSIDE / 2, self.BOARD_INSIDE / 2))
        total = sum(abs(x) for x in self.velocity)
        if total:
            self.ratios = [abs(x) / total for x in self.velocity]
        else:
            # Just in case we have a stationary object from the start
            # Shouldn't happen but just in case.
            self.ratios = [0.5, 0.5]
        self.symbol = symbol

    def tick(self, time: float, final: bool, collision_velocity=None):
        self.location = [
            self.location[0] + self.velocity[0] * time,
            self.location[1] + self.velocity[1] * time
        ]

        if collision_velocity:
            self.velocity = collision_velocity

        if final:
            if not self.moving:
                return
            if self.velocity[0] < 0:
                temp = min(0, self.velocity[0] - self.deceleration * self.ratios[0])
                assert abs(temp) <= abs(self.velocity[0])
                self.velocity[0] = temp
            else:
                temp = max(0, self.velocity[0] + self.deceleration * self.ratios[0])
                assert abs(temp) <= abs(self.velocity[0])
                self.velocity[0] = temp

            if self.velocity[1] < 0:
                temp = min(0, self.velocity[1] - self.deceleration * self.ratios[1])
                assert abs(temp) <= abs(self.velocity[1])
                self.velocity[1] = temp
            else:
                temp = max(0, self.velocity[1] + self.deceleration * self.ratios[1])
                assert abs(temp) <= abs(self.velocity[1])
                self.velocity[1] = temp

    def get_collisions(self, others: List[MovingObject], max_duration=1.0):
        new_loc = [
            self.location[0] + self.velocity[0],
            self.location[1] + self.velocity[1]
        ]
        collisions = []

        if new_loc[0] < 0 and abs(self.location[0] / self.velocity[0]) < max_duration:
            collisions.append(("LEFT_WALL", abs(self.location[0] / self.velocity[0]), self))

        if new_loc[0] > self.BOARD_INSIDE and abs(
                (self.BOARD_INSIDE - self.location[0]) / self.velocity[0]) < max_duration:
            collisions.append(("RIGHT_WALL", abs((self.BOARD_INSIDE - self.location[0]) / self.velocity[0]), self))

        if new_loc[1] < 0 and abs(self.location[1] / self.velocity[1]) < max_duration:
            collisions.append(("TOP_WALL", abs(self.location[1] / self.velocity[1]), self))

        if new_loc[1] > self.BOARD_INSIDE and abs(
                (self.BOARD_INSIDE - self.location[1]) / self.velocity[1]) < max_duration:
            collisions.append(("BOTTOM_WALL", abs((self.BOARD_INSIDE - self.location[1]) / self.velocity[1]), self))

        for other in others:
            if not self.moving and not other.moving:
                continue
            collision_time = collision(self.location, self.velocity, other.location, other.velocity)
            if 0 < collision_time < max_duration:
                collisions.append((other, collision_time, self))

        return collisions

    @property
    def position(self):
        return [int(round(self.location[0])) + 1, int(round(self.location[1])) + 1]

    @property
    def moving(self):
        return bool(self.velocity[0] or self.velocity[1])


def print_board(board):
    for y in range(BOARD_WIDTH):
        for x in range(BOARD_WIDTH):
            print(board[x + y * BOARD_WIDTH], end=" ")
        print()


def game():
    clear_board(board)

    tba_objects = []
    # the file should have one line per object space separated with three fields:
    # "how many game ticks to wait since last object before starting" "starting angle" "symbol"
    with open("tba_objects.txt") as f:
        for line in f:
            temp = line.strip().split()
            temp[0] = int(temp[0])
            temp[1] = int(temp[1])
            tba_objects.append(temp)

    objects = []
    while tba_objects[0][0] == 0:
        _, angle, symbol = tba_objects.pop(0)
        objects.append(MovingObject((math.cos(angle) * 5, -math.sin(angle) * 5), symbol))

    while True:
        start = datetime.now()
        print_board(board)

        tick_so_far = 0
        while tick_so_far < 1:
            col_vels = [None for _ in range(len(objects))]
            all_collision = []
            for i in range(len(objects)):
                ob = objects[i]
                collisions = ob.get_collisions(objects[i + 1:], 1 - tick_so_far)
                all_collision.extend(collisions)
            dur = 1 - tick_so_far
            if all_collision:
                # Allows processing simultaneously happening collisions and ensures that there are maximum of
                # 100 events in one game tick.
                # One problem with this method is that if same object has two collisions rapidly one of them can
                # be ignored, this will make the simulation not accurate but there should be no real problems
                # since even if the object ends up outside of the board and inside the wall it will be constantly
                # colliding with the wall until it is outside of the wall
                first_collision_time = math.ceil(min(all_collision, key=lambda x: x[1])[1] * 100) / 100
                used_collisions = []
                dur = first_collision_time
                for coll in all_collision:
                    if coll[1] <= first_collision_time:
                        used_collisions.append(coll)

                for coll in used_collisions:
                    other_object, _, collider = coll
                    idx = objects.index(collider)
                    if other_object == "LEFT_WALL" or other_object == "RIGHT_WALL":
                        col_vels[idx] = [-collider.velocity[0], collider.velocity[1]]
                    elif other_object == "TOP_WALL" or other_object == "BOTTOM_WALL":
                        col_vels[idx] = [collider.velocity[0], -collider.velocity[1]]
                    else:
                        other_idx = objects.index(other_object)
                        if other_object.moving and collider.moving:
                            # I'm not sure if this is 100% accurate way for two round objects with same mass
                            # but seems close enough
                            temp = collider.ratios
                            collider.ratios = other_object.ratios
                            other_object.ratios = temp
                            total_velocity = (math.sqrt(collider.velocity[0] ** 2 + collider.velocity[1] ** 2) +
                                              math.sqrt(
                                                  other_object.velocity[0] ** 2 + other_object.velocity[1] ** 2)) / 2
                            col_vels[other_idx] = [
                                total_velocity * other_object.ratios[0] * (-1 if collider.velocity[0] < 0 else 1),
                                total_velocity * other_object.ratios[1] * (-1 if collider.velocity[1] < 0 else 1)]
                            col_vels[idx] = [
                                total_velocity * collider.ratios[0] * (-1 if other_object.velocity[0] < 0 else 1),
                                total_velocity * collider.ratios[1] * (-1 if other_object.velocity[1] < 0 else 1)]
                        # Stationary objects remain stationary because I just want them to
                        else:
                            if collider.moving:
                                col_vels[idx] = [-collider.velocity[0], -collider.velocity[1]]
                            else:
                                col_vels[other_idx] = [-other_object.velocity[0], -other_object.velocity[1]]

            for ob, col_vel in zip(objects, col_vels):
                ob.tick(dur, not (tick_so_far + dur < 1), col_vel)

            if all_collision:
                for coll in used_collisions:
                    other_object, _, collider = coll
                    if other_object == "LEFT_WALL":
                        board[0 + int(collider.location[1] + 1.5) * BOARD_WIDTH] = collider.symbol
                    elif other_object == "RIGHT_WALL":
                        board[BOARD_WIDTH - 1 + int(collider.location[1] + 1.5) * BOARD_WIDTH] = collider.symbol
                    elif other_object == "TOP_WALL":
                        board[0 + int(collider.location[0] + 1.5)] = collider.symbol
                    elif other_object == "BOTTOM_WALL":
                        board[(BOARD_WIDTH - 1) * BOARD_WIDTH + int(collider.location[0] + 1.5)] = collider.symbol
            tick_so_far += dur

        clear_board(board)

        if tba_objects and tba_objects[0][0] == 0:
            _, angle, symbol = tba_objects.pop(0)
            objects.append(MovingObject((math.cos(angle) * 5, -math.sin(angle) * 5), symbol))

        for ob in objects:
            board[ob.position[0] + ob.position[1] * BOARD_WIDTH] = ob.symbol

        if not any(x.moving for x in objects) and not tba_objects:
            break

        if tba_objects:
            tba_objects[0][0] -= 1
        stop = datetime.now()

        sleep(0.5 - (stop - start).microseconds / 1000000)


def clear_board(board):
    for y in range(BOARD_WIDTH):
        for x in range(BOARD_WIDTH):
            if y != 0 and y != BOARD_WIDTH - 1 and x != 0 and x != BOARD_WIDTH - 1:
                board[x + y * BOARD_WIDTH] = "."


if __name__ == '__main__':
    game()

File: src/game.rs
pub const BOARD_WIDTH: usize = 16;
const BOARD_SIZE: usize = BOARD_WIDTH * BOARD_WIDTH;

const BOARD_INSIDE: usize = BOARD_WIDTH - 3;

pub static mut BOARD: [char; BOARD_SIZE] = ['X'; BOARD_SIZE];

const DECELERATION: f32 = -0.08f32;

pub const MAXIMUM_OBJECTS: usize = 10;
const MAX_COLLISIONS_PER_OBJECT: usize = 10;
const MAXIMUM_COLLISIONS: usize = MAXIMUM_OBJECTS * MAX_COLLISIONS_PER_OBJECT;

pub fn clear_board() {
    for y in 1..(BOARD_WIDTH - 1) {
        for x in 1..(BOARD_WIDTH - 1) {
            unsafe {
                BOARD[x + y * BOARD_WIDTH] = '.';
            }
        }
    }
}

pub fn print_board() {
    for y in 0..(BOARD_WIDTH) {
        for x in 0..(BOARD_WIDTH) {
            //unsafe {
            //}
        }
    }
}


fn solve_quadratic(a: f32, b: f32, c: f32) -> f32 {
    let determinant: f32 = b * b - 4f32 * a * c;
    // Since we only care about collisions that happen between t=[0, 1)
    // it is ok to return -1 here
    if determinant < 0f32 {
        -1f32
    } else {
        (-b - fast_sqrt(determinant)) / (2f32 * a)
    }
}

fn find_collision_times(start1: Vector, v1: Vector, start2: Vector, v2: Vector, radius: f32) -> f32 {
    let a: f32 = pow2(v1.x - v2.x) + pow2(v1.y - v2.y);

    let b: f32 = -2f32 * (v1.x - v2.x) * (start2.x - start1.x)
        - 2f32 * (v1.y - v2.y) * (start2.y - start1.y);

    let c: f32 = pow2(start1.x - start2.x) +
        pow2(start1.y - start2.y) - pow2(radius);

    if abs(a) < f32::EPSILON {
        return -c / b;
    }
    if abs(a) < f32::EPSILON && abs(b) < f32::EPSILON {
        return -1f32;
    }
    solve_quadratic(a, b, c)
}

#[derive(Debug, Copy, Clone)]
pub struct Vector {
    pub(crate) x: f32,
    pub(crate) y: f32,
}


#[derive(Debug, Clone)]
pub struct MovingObject {
    velocity: Vector,
    location: Vector,
    ratios: Vector,
    pub(crate) symbol: char,
    age: usize,
}

impl MovingObject {
    pub fn new(starting_location: Vector, starting_velocity: Vector, symbol: char) -> MovingObject {
        let location = starting_location;
        let sum = abs(starting_velocity.x) + abs(starting_velocity.y);
        if sum != 0f32 {
            let temp = Vector { x: abs(starting_velocity.x) / sum, y: abs(starting_velocity.y) / sum };
            MovingObject {
                velocity: starting_velocity,
                location,
                symbol,
                ratios: temp,
                age: 0,
            }
        } else {
            MovingObject {
                velocity: starting_velocity,
                location,
                symbol,
                ratios: Vector { x: 0.5f32 / sum, y: 0.5f32 },
                age: 0,
            }
        }
    }

    fn tick(&mut self, time: f32, final_: bool, collision_velocity: Option<Vector>) {
        if !self.moving() {
            return;
        }
        self.location = Vector {
            x: self.location.x + self.velocity.x * time,
            y: self.location.y + self.velocity.y * time,
        };

        // println!("{} {}", self.velocity.x, self.velocity.y);
        match collision_velocity {
            Some(x) => self.velocity = x,
            None => (),
        }

        if final_ {
            let x_vel;
            let y_vel;
            if self.velocity.x < 0f32 {
                x_vel = min(0f32, self.velocity.x - DECELERATION * self.ratios.x)
            } else {
                x_vel = max(0f32, self.velocity.x + DECELERATION * self.ratios.x)
            }
            if self.velocity.y < 0f32 {
                y_vel = min(0f32, self.velocity.y - DECELERATION * self.ratios.y)
            } else {
                y_vel = max(0f32, self.velocity.y + DECELERATION * self.ratios.y)
            }
            self.velocity = Vector { x: x_vel, y: y_vel };
        }
    }

    fn get_collisions(&self,
                      others: &[Option<MovingObject>; MAXIMUM_OBJECTS],
                      offset: usize,
                      max_duration: f32, ) -> [Option<(usize, f32, usize)>; MAX_COLLISIONS_PER_OBJECT] {
        let new_loc = Vector {
            x: self.location.x + self.velocity.x,
            y: self.location.y + self.velocity.y,
        };
        // println!("{} {}", self.velocity.x, self.velocity.y);

        let mut num_collisions: usize = 0;
        let mut collisions: [Option<(usize, f32, usize)>; MAX_COLLISIONS_PER_OBJECT] = [None; MAX_COLLISIONS_PER_OBJECT];

        if new_loc.x < 0f32 && abs(self.location.x / self.velocity.x) < max_duration {
            collisions[num_collisions] = Some((127, abs(self.location.x / self.velocity.x), offset));
            num_collisions += 1;
        }
        if new_loc.x > BOARD_INSIDE as f32 &&
            abs((BOARD_INSIDE as f32 - self.location.x) / self.velocity.x) < max_duration {
            collisions[num_collisions] = Some((126, abs((BOARD_INSIDE as f32 - self.location.x) / self.velocity.x), offset));
            num_collisions += 1;
        }
        if new_loc.y < 0f32 && abs(self.location.y / self.velocity.y) < max_duration {
            collisions[num_collisions] = Some((125, abs(self.location.y / self.velocity.y), offset));
            num_collisions += 1;
        }
        if new_loc.y > BOARD_INSIDE as f32 &&
            abs((BOARD_INSIDE as f32 - self.location.y) / self.velocity.y) < max_duration {
            collisions[num_collisions] = Some((124, abs((BOARD_INSIDE as f32 - self.location.y) / self.velocity.y), offset));
            num_collisions += 1;
        }

        for i in offset + 1..MAXIMUM_OBJECTS {
            let other = match &others[i] {
                Some(o) => o,
                None => continue
            };
            if !self.moving() && !other.moving() {
                continue;
            }
            let collision_time = find_collision_times(
                self.location, self.velocity, other.location, other.velocity, 1f32,
            );
            if 0f32 < collision_time && collision_time < max_duration {
                collisions[num_collisions] = Some((i, collision_time, offset));
                num_collisions += 1;
            }
        }
        return collisions;
    }

    pub fn position(&self) -> (usize, usize) {
        ((self.location.x + 1.5) as usize, (self.location.y + 1.5) as usize)
    }

    pub fn moving(&self) -> bool {
        return self.velocity.x != 0f32 || self.velocity.y != 0f32;
    }
    pub fn clear_symbol(&mut self)  {
        self.symbol = '.';
    }
    pub fn add_age(&mut self)  {
        self.age = self.age + 1;
    }
    pub fn get_age(&self) -> usize {
        self.age
    }
}

pub fn game_tick(objects: &mut [Option<MovingObject>; 10], number_of_objects: usize) {
    let mut tick_so_far = 0f32;
    while tick_so_far < 1f32 {
        let mut collision_velocities: [Option<Vector>; MAXIMUM_OBJECTS] = [None; MAXIMUM_OBJECTS];
        let mut all_collisions: [Option<(usize, f32, usize)>; MAXIMUM_COLLISIONS] = [None; MAXIMUM_COLLISIONS];
        let mut total_collisions: usize = 0;
        for i in 0..number_of_objects {
            let ob = match &objects[i] {
                Some(o) => o,
                None => continue,
            };
            let temp = ob.get_collisions(objects, i, 1f32 - tick_so_far);
            for i in 0..MAX_COLLISIONS_PER_OBJECT {
                match temp[i] {
                    Some(_) => (),
                    None => break
                };
                all_collisions[total_collisions] = temp[i];
                total_collisions += 1;
            }
        }
        let mut duration = 1f32 - tick_so_far;

        let mut num_used_collisions = 0;
        let mut used_collisions: [Option<(usize, f32, usize)>; MAXIMUM_COLLISIONS] = [None; MAXIMUM_COLLISIONS];

        if total_collisions != 0 {
            let mut first_collision = 1f32;

            for collision in &all_collisions {
                let time = match collision {
                    Some(t) => t.1,
                    None => break
                };
                first_collision = min(first_collision, time);
            }

            first_collision = (first_collision * 100f32 + 1f32) / 100f32;
            duration = first_collision;

            for collision in &all_collisions {
                let time = match collision {
                    Some(t) => t.1,
                    None => break
                };
                if time <= first_collision {
                    used_collisions[num_used_collisions] = *collision;
                    num_used_collisions += 1;
                }
            }
        }
        for collision in &used_collisions {
            let (other_idx, first_idx) = match collision {
                Some(o) => (o.0, o.2),
                None => break,
            };
            let mut first_collider = match objects[first_idx].clone() {
                Some(o) => o,
                None => panic!()
            };

            if other_idx == 126 || other_idx == 127 {
                collision_velocities[first_idx] = Some(Vector {
                    x: -first_collider.velocity.x,
                    y: first_collider.velocity.y,
                });
            } else if other_idx == 124 || other_idx == 125 {
                collision_velocities[first_idx] = Some(Vector {
                    x: first_collider.velocity.x,
                    y: -first_collider.velocity.y,
                });
            } else {
                let other_collider = &mut match objects[other_idx].clone() {
                    Some(o) => o,
                    None => panic!()
                };
                if other_collider.moving() && first_collider.moving() {
                    let temp = first_collider.ratios;
                    first_collider.ratios = other_collider.ratios;
                    other_collider.ratios = temp;

                    let total_velocity = (
                        fast_sqrt(pow2(first_collider.velocity.x) + pow2(first_collider.velocity.y)) +
                            fast_sqrt(pow2(other_collider.velocity.x) + pow2(other_collider.velocity.y))
                    ) / 2f32;

                    collision_velocities[other_idx] = Some(Vector {
                        x: total_velocity * other_collider.ratios.x * (if first_collider.velocity.x < 0f32 { -1f32 } else { 1f32 }),
                        y: total_velocity * other_collider.ratios.y * (if first_collider.velocity.y < 0f32 { -1f32 } else { 1f32 }),
                    });

                    collision_velocities[first_idx] = Some(Vector {
                        x: total_velocity * first_collider.ratios.x * (if other_collider.velocity.x < 0f32 { -1f32 } else { 1f32 }),
                        y: total_velocity * first_collider.ratios.y * (if other_collider.velocity.y < 0f32 { -1f32 } else { 1f32 }),
                    });
                } else if first_collider.moving() {
                    collision_velocities[first_idx] = Some(Vector {
                        x: -first_collider.velocity.x,
                        y: -first_collider.velocity.y,
                    });
                } else {
                    collision_velocities[other_idx] = Some(Vector {
                        x: -other_collider.velocity.x,
                        y: -other_collider.velocity.y,
                    });
                }
            }
        }
        for i in 0..number_of_objects {
            let temp = &mut match objects[i].clone() {
                Some(o) => o,
                None => continue
            };
            temp.tick(duration,
                      !(tick_so_far + duration < 1f32),
                      collision_velocities[i]);

        }

        if num_used_collisions != 0 {
            for collision in &used_collisions {
                let (other_idx, first_idx) = match collision {
                    Some(o) => (o.0, o.2),
                    None => break,
                };
                let object = match &objects[first_idx] {
                    Some(o) => o,
                    None => panic!()
                };
                if other_idx == 127 {
                    unsafe {
                        BOARD[0 + (object.location.y + 1.5f32) as usize * BOARD_WIDTH] = object.symbol;
                    }
                } else if other_idx == 126 {
                    unsafe {
                        BOARD[BOARD_WIDTH - 1 + (object.location.y + 1.5f32) as usize * BOARD_WIDTH] = object.symbol;
                    }
                } else if other_idx == 125 {
                    unsafe {
                        BOARD[(object.location.x + 1.5f32) as usize + 0] = object.symbol;
                    }
                } else if other_idx == 124 {
                    unsafe {
                        BOARD[(object.location.x + 1.5f32) as usize + (BOARD_WIDTH - 1) * BOARD_WIDTH] = object.symbol;
                    }
                }
            }
        }
        tick_so_far += duration;
    }
}


#[repr(C)]
union MyUnion {
    f1: u32,
    f2: f32,
}

pub fn fast_sqrt(num: f32) -> f32 {
    let x2 = num * 0.5f32;
    let threehalfs = 1.5f32;

    let mut conv = MyUnion { f2: num };
    unsafe {
        conv.f1 = 0x5f3759df - (conv.f1 >> 1);

        conv.f2 *= threehalfs - (x2 * conv.f2 * conv.f2);
        return 1f32 / conv.f2;
    }
}

pub fn abs(num: f32) -> f32 {
    let mut conv = MyUnion { f2: num };
    unsafe {
        conv.f1 &= 0x7fffffff;
        return conv.f2;
    }
}

pub fn pow2(num: f32) -> f32 {
    num * num
}

fn min(first: f32, second: f32) -> f32 {
    if first < second { first } else { second }
}

fn max(first: f32, second: f32) -> f32 {
    if first > second { first } else { second }
}